sw/device/lib/crypto/impl/hash.c

```diff
@@ -9,125 +9,57 @@
-  // As per the `hardened_memcpy()` documentation, it is OK to cast to
-  // `uint32_t *` here as long as `state` is word-aligned, which it must be
-  // because all its fields are.
-  hardened_memcpy(ctx->data, (uint32_t *)state,
-                  sizeof(sha384_state_t) / sizeof(uint32_t));
-}
+    sizeof(otcrypto_hash_context_t) >= sizeof(hmac_ctx_t),
+    "`otcrypto_hash_context_t` must be big enough to hold `hmac_ctx_t`.");
 
-/**
- * Restore a SHA-384 state from a generic hash context.
- *
- * @param ctx Generic hash context to restore from.
- * @param[out] state Destination SHA-384 context object.
- */
-static void sha384_state_restore(const otcrypto_hash_context_t *restrict ctx,
-                                 sha384_state_t *restrict state) {
-  // As per the `hardened_memcpy()` documentation, it is OK to cast to
-  // `uint32_t *` here as long as `state` is word-aligned, which it must be
-  // because all its fields are.
-  hardened_memcpy((uint32_t *)state, ctx->data,
-                  sizeof(sha384_state_t) / sizeof(uint32_t));
-}
+// Ensure that HMAC driver struct is suitable for `hardened_memcpy()`.
+static_assert(sizeof(hmac_ctx_t) % sizeof(uint32_t) == 0,
+              "Size of `hmac_ctx_t` must be a multiple of the word size for "
+              "`hardened_memcpy()`");
 
 /**
- * Save a SHA-512 state to a generic hash context.
+ * Save the internal HMAC driver context to a generic hash context.
  *
  * @param[out] ctx Generic hash context to copy to.
- * @param state SHA-512 context object.
+ * @param hmac_ctx The internal context object from HMAC driver.
  */
-static void sha512_state_save(otcrypto_hash_context_t *restrict ctx,
-                              const sha512_state_t *restrict state) {
-  // As per the `hardened_memcpy()` documentation, it is OK to cast to
-  // `uint32_t *` here as long as `state` is word-aligned, which it must be
-  // because all its fields are.
-  hardened_memcpy(ctx->data, (uint32_t *)state,
-                  sizeof(sha512_state_t) / sizeof(uint32_t));
+static void hmac_ctx_save(otcrypto_hash_context_t *restrict ctx,
+                          const hmac_ctx_t *restrict hmac_ctx) {
+  hardened_memcpy(ctx->data, (uint32_t *)hmac_ctx,
+                  sizeof(hmac_ctx_t) / sizeof(uint32_t));
 }
 
 /**
- * Restore a SHA-512 state from a generic hash context.
+ * Restore an internal HMAC driver context from a generic hash context.
  *
  * @param ctx Generic hash context to restore from.
- * @param[out] state Destination SHA-512 context object.
+ * @param[out] hmac_ctx Destination HMAC driver context object.
  */
-static void sha512_state_restore(const otcrypto_hash_context_t *restrict ctx,
-                                 sha512_state_t *restrict state) {
-  // As per the `hardened_memcpy()` documentation, it is OK to cast to
-  // `uint32_t *` here as long as `state` is word-aligned, which it must be
-  // because all its fields are.
```

Comment:
I'd leave this comment in; that cast is normally unsafe, so it's good to record the justification for why it's OK here. Same for the comment in `save` above.

Reply:
Added the same comment to all four functions (as there are duplicates of the same functions in `mac.c` and `hash.c`).